feat: Relic order claims and returns

starknetdev opened the pull request:
According to #175 I am beginning the implementation for claims of relics by order and for the return of relics for unsettling realms.

Before implementing it in full, I wanted to write some 'pseudo code' steps to explain my thought process on how to tackle handling multiple relics for owners, it is similar to how I've approached token storage in guildly (it has some drawbacks with order of complexity O(n)):

- storage variable for array of relics by owner (plus a count store) - `(owner_token_id, index) -> (relic_id)`
- on relic set, find relic index of old owner and set to `Uint256(0,0)`(this requires looping through entire array to find relic index), add relic to new owner store
- looping through arrays when setting/returning relics will also loop through previously removed ones (the count of relics needs to remain the same when removing a relic from a previous owner)

this requires increased steps and computation can become large when relics are changed around a bit. any thoughts on the above implementation? if OK I will proceed with these params

ponderingdemocritus commented on contracts/settling_game/modules/relics/Relics.cairo:
```diff
@@ -148,3 +248,51 @@ func get_current_relic_holder{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, ra
+
+    assert relics[index] = relic;
+
+    loop_get_relics(
```
move to return, and remove identifiers for cleanliness

ponderingdemocritus reviewed (commented):
Looks good. Have left some comments. Can't see issue with logic.

Could you also add natspec on each function

milancermak commented on contracts/settling_game/modules/relics/Relics.cairo:
```diff
@@ -97,31 +110,118 @@ func set_relic_holder{pedersen_ptr: HashBuiltin*, syscall_ptr: felt*, range_chec
+
+    if (winner_realm_data.order == relic_owner_data.order) {
+        _set_relic_holder(relic_id, relic_id);
+        return ();
```
Why the early return?

starknetdev replied:
Good spot, needs to be removed so loops through all relics. Have also added to test to verify it works for returning multiple relics to same order.

milancermak commented on contracts/settling_game/modules/relics/Relics.cairo:
```diff
@@ -97,31 +110,118 @@ func set_relic_holder{pedersen_ptr: HashBuiltin*, syscall_ptr: felt*, range_chec
+    }
+    // _set_relic_holder(relic_id, relic_id);
+
+    loop_claim_order_relic(index + 1, realms_address, winner_realm_data, loser_token_id, loser_relics_len);
```
Should be `return loop_claim_order_relic...`

milancermak commented on contracts/settling_game/utils/general.cairo:
```diff
@@ -245,3 +246,21 @@ func sum_values_by_key_loop{range_check_ptr}(
+        return (index=arr_index);
+    }
+
+    let (arr_index) = find_uint256_value(arr_index + 1, arr_len, arr, value);
```
Should be `return find...` here as well

milancermak commented on contracts/settling_game/modules/relics/library.cairo:
```diff
@@ -32,4 +34,20 @@ namespace Relics {
+    // @param relics: list of relics a realm owns, Uint256
+    // @param relic_id: id of relic to find in list
+    // @return relic_index: returns index of relic in list
+    func _get_relic_index{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
```
IMO this is a needless function. It only wraps a call to another function.

milancermak commented on contracts/settling_game/modules/relics/Relics.cairo:
```diff
@@ -97,31 +110,118 @@ func set_relic_holder{pedersen_ptr: HashBuiltin*, syscall_ptr: felt*, range_chec
     storage_relic_holder.write(relic_id, owner_token_id);
 
+    // If old owner exists, update array
+    let (check_exists) = uint256_lt(Uint256(0,0), old_owner_token_id);
```
This will never be true, will it?

starknetdev replied:
It will be true when the owner has previously been set from a battle.

Example (-> = wins battle, Realm A and C same order):

Realm B -> Realm A (A's relic set to realm B, `storage_relic_holder(A) = B`)
Realm C -> Realm B (A's relic returned, B's relics updated to 0. `storage_relic_holder(A) = B` not uint(0))

milancermak replied:
Ah, my bad, I read it wrong. Isn't the issue tho, contrary to my previous message, that it will always be true?

In the simulation you describe above, why is B's relic set to 0? Doesn't C capture it? Where in the code does this setting to 0 happen? Is it the `owned_relics.write(old_owner_token_id, relic_index, Uint256(0, 0));` call 2 lines below? If so, why is it a different storage_var?

Sorry, it looks like I'm not aware of some piece of here 🥶

starknetdev replied:
C won't capture it because A and C are in the same order, so that relic needs to be given back to A rather than captured. After those 2 battles:

A has relic A.
B has no relics.
C has relics C and B.

Yes every index in the `owned_relics` array of for B needs to be set to 0. B was previously holding 2, it's own and A's, we are updating the array to keep track that it no longer holds A's. This array storage is needed because B could have n number of other realms relics.

starknetdev replied:
`let (check_exists) = uint256_lt(Uint256(0,0), old_owner_token_id);` will be false at the first loss of each realm. There is no initialization of relic owners, so when `storage_relic_holder = uint(0)` then the realm holds it own relic and hasn't lost it yet

milancermak replied:
The `old_owner_token_id` is never going to be 0 (*), is it not? You're setting zeros in the `owned_relics` storage var, but comparing for zero a val read from `storage_relic_holder`. And the only place `storage_relic_holder` is being written to is in this function, with the new `owner_token_id` value.

(*) -- actually, **initally** the whole `storage_relic_holder` is going to be zero, because all the values are uninitialized. 

> every index in the owned_relics array of for B needs to be set to 0

You're only setting a single index tho, not the whole array.

starknetdev replied:
the `loop_claim_order_relic` calls `_set_relic_holder` for each relic, so we set the single index but for each relic for the all relics B owns (in the case above)

milancermak reviewed (commented):
Left a couple of comments inline.

One comment that applies more broadly: looping function should end with a `return func()` call. I'm 80% certain the Cairo compiler won't do tail call optimization otherwise.

starknetdev commented:
Comments have been implemented in latest commit!

milancermak commented:
Please don't merge this yet.

milancermak commented on contracts/settling_game/modules/relics/Relics.cairo:
```diff
@@ -97,31 +112,145 @@ func set_relic_holder{pedersen_ptr: HashBuiltin*, syscall_ptr: felt*, range_chec
+
+    let (winner_realm_data: RealmData) = IRealms.fetch_realm_data(realms_address, winner_token_id);
+
+    let (loser_relics_len) = owned_relics_count.read(loser_token_id);
```
This value, the number of relics owned by the loser, is never decremented. Implies that the `owned_relics` array is never purged of 0 values. It will only grow, eventually looking like e.g. `[0, 0, 0, ....., 782]` if I'm not mistaken.

starknetdev replied:
Yes this is the current case, something I mentioned at the start of the PR. I can't see a way around this currently, as we need to keep track of all relics owned where a relic will be removed at a random index in an array.

E.g removing index `1` from `[1,2,3]` leaves `[1,0,3]`. We cannot decrement the count as we would then only be looping on `[1,0]` where `3` still exists in the array.

Any suggestions of how we can get around this 🤔

starknetdev replied:
Now I think of it perhaps just creating a new array. Looping over the original and only pushing the value when it is bigger than 0. Then storing this over the original array storage variable.

milancermak replied:
This is a problem for an infinite game. We can't merge it as is.

Maybe using some different data model would help? Or doing some bit packing? I don't have a clear solution in mind. I would also be in favor of altering the whole mechanic TBH. We have to work within the constraints of the system.

ponderingdemocritus replied:
We could move the data off-chain - so the victor needs to enter the Relics they want to capture back

starknetdev commented:
As per our conversation I have implemented new logic to decrement the array.

Quite a few changes in `Relics.cairo` since the last commit, so please review carefully. I have thought about the most efficient way (in the circumstances) to manage the array, narrowing it down to only one looping function `loop_remove_relic`. As you will see it takes 2 index parameters which allows iteration through the old array, while also create a squashed one in the same loop.

I have removed 'get_relic_index' and 'get_relics' as everything is now being handled in one loop.

starknetdev commented on contracts/settling_game/modules/relics/Relics.cairo:
```diff
@@ -97,36 +111,165 @@ func set_relic_holder{pedersen_ptr: HashBuiltin*, syscall_ptr: felt*, range_chec
+// @param relics_len: length of old owned relic array
+// @param owner_token_id: realm id of old relic owner
+// @param relic_id: relic id, original relic owner realm id
+func loop_remove_relic{pedersen_ptr: HashBuiltin*, syscall_ptr: felt*, range_check_ptr}(
```
Here's an example of what I'm going for:

Let's remove Relic `2` from `[1, 2, 3, 4]`. Position in the array = `<x>`

`old_index = 0`, `new_index = 0`. 
`[<1>, 2, 3, 4]` - array becomes `[<1>, 2, 3, 4]`, increment both indexes by 1
               
`old_index = 1`, `new_index = 1`. 
`[1, <2>, 3, 4]` - we set array to `old_index + 1`, array becomes `[1, <3>, 3, 4]`, increment old_index by 2, new_index 1

`old_index = 3`, `new_index = 2`. 
`[1, 3, 3, <4>]` - array becomes `[1, 3, <4>, 4]`, increment both indexes by 1

`old_index = 4`, `new_index = 3`. 
`[1, 3, 4, 4]` stop as we have hit `relic_count`, reduce count so array becomes `[1, 3, 4]`

1. As above these indexes keep track of where the value is, then increment differently to allow shifting down of values.
2. We can't return as we still need to shift down the values after it has been found.
3. In the case where the relics array is `[1]` then with the logic above `old_index = 2` which is above the `relic_count` so I needed to add that for these cases (where the relic to be removed is at the end of the array).

starknetdev commented:
As @milancermak suggested I have now implemented in-memory array of relics to be removed before the loop to remove them. If my approach as stated in the previous comment is OK then this is should be good to go.

milancermak commented on tests/protostar/settling_game/relics/test_relics.cairo:
```diff
@@ -37,3 +38,19 @@ func test_current_relic_holder_stolen{
 }
+
+@external
+func test_relic_index{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() {
```
No need for this test anymore, right?

milancermak commented on contracts/settling_game/modules/relics/Relics.cairo:
```diff
@@ -97,36 +130,226 @@ func set_relic_holder{pedersen_ptr: HashBuiltin*, syscall_ptr: felt*, range_chec
+    loser_relics_len: felt,
+    remove_relics_len: felt,
+    remove_relics: Uint256*,
+) -> (remove_relics_len: felt) {
```
Please rename the return value. It has the same name as a passed in argument which is weird.

milancermak commented on contracts/settling_game/modules/relics/Relics.cairo:
```diff
@@ -97,36 +130,226 @@ func set_relic_holder{pedersen_ptr: HashBuiltin*, syscall_ptr: felt*, range_chec
+    );
+
+    loop_remove_relic(0, 0, loser_relics_len, old_owner_token_id, remove_relics_len, remove_relics);
+    owned_relics_len.write(old_owner_token_id, loser_relics_len - remove_relics_len);
```
Have you tested this? Does this work as intended? What if the first branch above matches (line 112 and below)? Then `loser_relics_len` is already off by one (line 114), no?

Furthermore, `loop_claim_order_relic` calls `_set_relic_holder` which sets the `owned_relics_len` value. Then `loop_remove_relic` updates the `owned_relics` array without updating the `_len` (IMO this should be done in lockstep). And finally, the `owned_relics_len` is once more updated here.

milancermak commented on contracts/settling_game/modules/relics/Relics.cairo:
```diff
@@ -97,36 +130,226 @@ func set_relic_holder{pedersen_ptr: HashBuiltin*, syscall_ptr: felt*, range_chec
 // SETTERS
 // -----------------------------------
 
+// @notice get index of relic in the array of relics held
```
The comment doesn't reflect what the function does.

milancermak commented on contracts/settling_game/modules/relics/Relics.cairo:
```diff
@@ -97,36 +130,226 @@ func set_relic_holder{pedersen_ptr: HashBuiltin*, syscall_ptr: felt*, range_chec
+    if (winner_realm_data.order == relic_owner_data.order) {
+        _set_relic_holder(relic_id, relic_id);
+        assert remove_relics[index] = relic_id;
+        loop_claim_order_relic(
```
Missing a `return` which will probably allow to remove all the rebinding.

milancermak commented on contracts/settling_game/modules/relics/Relics.cairo:
```diff
@@ -97,36 +130,226 @@ func set_relic_holder{pedersen_ptr: HashBuiltin*, syscall_ptr: felt*, range_chec
+
+    if (winner_realm_data.order == relic_owner_data.order) {
+        _set_relic_holder(relic_id, relic_id);
+        assert remove_relics[index] = relic_id;
```
Why is this using `index`? Shouldn't it be `remove_relics_len`?

milancermak reviewed (commented):
Left some comments inline. It's quite difficult to follow the flow, but I don't think the PR is ready, but could also be that I'm stupid. Awaiting clarifications.

starknetdev commented:
Review comments have been implemented.

@milancermak `remove_relics_len` is now decremented from `owned_relics_len` within `loop_remove_relic`.

milancermak commented on contracts/settling_game/modules/relics/Relics.cairo:
```diff
@@ -97,36 +129,220 @@ func set_relic_holder{pedersen_ptr: HashBuiltin*, syscall_ptr: felt*, range_chec
+        0, realms_address, winner_realm_data, loser_token_id, loser_relics_len, 0, remove_relics
+    );
+
+    return loop_remove_relic(0, 0, loser_relics_len, old_owner_token_id, remove_relics_len, remove_relics);
```
I think this shouldn't be `old_owner_token_id` but rather `loser_token_id`.

milancermak commented on contracts/settling_game/modules/relics/Relics.cairo:
```diff
@@ -97,36 +129,220 @@ func set_relic_holder{pedersen_ptr: HashBuiltin*, syscall_ptr: felt*, range_chec
+    let (loser_relics_len) = owned_relics_len.read(loser_token_id);
+
+    let (remove_relics_len) = loop_claim_order_relic(
+        0, realms_address, winner_realm_data, loser_token_id, loser_relics_len, 0, remove_relics
```
This `remove_relics` is potentially already set to `[loser_token_id]` from line 112, so then an attempt to write into index 0 on line 178 would fail.

milancermak reviewed (changes requested):
I think there are still two bugs in there.

starknetdev commented:
Thanks @milancermak for checking carefully, you are quite right.

Implemented comments and added a repeat claim order in `test_relics_logic` to ensure the arrays are updating as they should.

Also re-added a crucial `Uint256` dependency in `general.cairo` as I got confused with Amarna.

ponderingdemocritus commented:
are we good to merge this?

milancermak commented on contracts/settling_game/modules/relics/Relics.cairo:
```diff
@@ -75,19 +88,39 @@ func set_relic_holder{pedersen_ptr: HashBuiltin*, syscall_ptr: felt*, range_chec
+        if (check_exists == TRUE) {
+            let (loser_relics_len) = owned_relics_len.read(loser_token_id);
+            assert remove_victory_relics[0] = loser_token_id;
+            loop_remove_relic(0, 0, loser_relics_len, loser_token_id, 1, remove_victory_relics);
```
I'm a bit uneasy about this part. Imagine this function getting called with `loser_relics_len` equal to 0. Then the `check` inside it will be `TRUE` and `0 - 1` (a "very big number" in Cairo) will get written in the storage var.

I think the scenario can happen. Say we have realms A, B & C, A & C in the same order and the game is in initial state (storage vars are zeroed). A fights B, wins. B fights A, wins. C fights B, wins. C unstakes (so relics are returned). A fights B, wins. I think in this scenario both checks will be TRUE (so we get to this part), but also `loser_relics_len` will be 0 because of the C vs. B fight (where B lost). Or do I have it wrong?

starknetdev replied:
`loser_relics_len` will not be 0 as `loop_return_relics` calls `_set_relic_holder` which increments B's count on the return. i.e we are taking into account that B has gained a relic when C unstaked (it was returned)

I've added a stress test for this, and read the relics length of 'Realm B' after the scenario asserting it is 0.

milancermak commented on contracts/settling_game/modules/relics/Relics.cairo:
```diff
@@ -97,36 +130,220 @@ func set_relic_holder{pedersen_ptr: HashBuiltin*, syscall_ptr: felt*, range_chec
+    alloc_locals;
+    // Only Settling
+    Module.only_approved();
+    let (relics_len) = owned_relics_len.read(realms_token_id);
```
Shouldn't this be reset to 0 for the `realms_token_id` at some point?

starknetdev replied:
`loop_return_relics` function builds the `remove_relics` array (being all the relics they own) and the count is reset `loop_remove_relic`. 

in `loop_remove_relic`:

`owned_relics_len.write(owner_token_id, relics_len - remove_relics_len);` where `remove_relics_len` is the same as `relics_len` in this case

milancermak reviewed (commented):
Left some more comments.

milancermak reviewed (approved):
Alright, let's merge this.
